File: erlang/strain/strain.erl
-module(strain).
-export([keep/2, discard/2]).

keep(F, List) ->
    keep(F, List, []).

keep(_, [], Acc) -> lists:reverse(Acc);
keep(F, [H|T], Acc) -> keep_if(F(H), F, [H|T], Acc).

keep_if(true, F, [H|T], Acc) -> keep(F, T, [H|Acc]);
keep_if(false, F, [_|T], Acc) -> keep(F, T, Acc).

discard(F, List) -> keep(fun(X) -> not F(X) end, List).

File: scala/strain/src/main/scala/Strain.scala
import scala.annotation.tailrec

object Strain {
  def keep[A](xs: Seq[A], f: A => Boolean): Seq[A] = {
    @tailrec
    def doKeep(xs: Seq[A], f: A => Boolean, acc: Seq[A]): Seq[A] = {
      xs match {
        case Seq() => acc
        case _ => {
          if (f(xs.head))
            doKeep(xs.tail, f, acc :+ xs.head)
          else
            doKeep(xs.tail, f, acc)
        }
      }

    }

    doKeep(xs, f, Seq[A]())
  }

  def discard[A](xs: Seq[A], f: A => Boolean): Seq[A] =
    keep(xs, (x: A) => !f(x))
}